sdk/go/pulumi/config/config_test.go

```diff
@@ -149,3 +151,180 @@ func TestConfig(t *testing.T) {
 	_, err = cfg.Try("missing")
 	assert.NotNil(t, err)
 }
+
+func TestSecretConfig(t *testing.T) {
+	ctx, err := pulumi.NewContext(context.Background(), pulumi.RunInfo{
+		Config: map[string]string{
+			"testpkg:sss":    "a string value",
+			"testpkg:bbb":    "true",
+			"testpkg:intint": "42",
+			"testpkg:fpfpfp": "99.963",
+			"testpkg:obj": `
+				{
+					"foo": {
+						"a": "1",
+						"b": "2"
+					},
+					"bar": "abc"
+				}
+			`,
+			"testpkg:malobj": "not_a_struct",
+		},
+	})
+	assert.Nil(t, err)
+
+	cfg := New(ctx, "testpkg")
+
+	fooMap := make(map[string]string)
+	fooMap["a"] = "1"
+	fooMap["b"] = "2"
+	expectedTestStruct := TestStruct{
+		Foo: fooMap,
+		Bar: "abc",
+	}
+
+	s1, err := cfg.TrySecret("sss")
+	s2 := cfg.RequireSecret("sss")
+	s3 := cfg.GetSecret("sss")
+	assert.Nil(t, err)
+
+	assert.True(t, s1.IsSecret())
+	assert.True(t, s2.IsSecret())
+	assert.True(t, s3.IsSecret())
+
+	errChan := make(chan error)
+	result := make(chan string)
+
+	pulumi.All(s1, s2, s3).ApplyT(func(v []interface{}) ([]interface{}, error) {
+		for _, val := range v {
+			if val == "a string value" {
+				result <- val.(string)
+			} else {
+				errChan <- errors.Errorf("Invalid result: %v", val)
+
+			}
+		}
+		return v, nil
+	})
+
+	for i := 0; i < 3; i++ {
+		select {
+		case err = <-errChan:
+			assert.Nil(t, err)
+			break
+		case r := <-result:
+			assert.Equal(t, "a string value", r)
+			break
+		}
+	}
```

Comment:
Is this style common in our Go tests?  Does it not work to just do the asserts in the `Apply`?

Reply:
I think in the past I've used a `WaitGroup` to accomplish this and either run the asserts in the `Apply` or leaked the resolved values to the outside and validated them there.